File: src/code-samples/langchain/langgraph-interrupts-hitl-stream.py
# :remove-start:
from typing import TypedDict

from langgraph.checkpoint.memory import InMemorySaver
from langgraph.graph import END, START, StateGraph
from langgraph.types import Command, interrupt


class StreamState(TypedDict):
    done: bool


def stream_node(state: StreamState):
    interrupt("wait for user")
    return {"done": True}


graph = (
    StateGraph(StreamState)
    .add_node("n", stream_node)
    .add_edge(START, "n")
    .add_edge("n", END)
    .compile(checkpointer=InMemorySaver())
)

config = {"configurable": {"thread_id": "stream-1"}}
initial_input: dict = {}


def display_streaming_content(content: str) -> None:
    pass


def get_user_input(interrupt_info: object) -> str:
    return "ok"
# :remove-end:

# :snippet-start: langgraph-interrupts-hitl-stream-py
from langchain.messages import AIMessageChunk
from langgraph.types import Command

for chunk in graph.stream(
    initial_input,
    stream_mode=["messages", "updates", "values"],
    subgraphs=True,
    config=config,
    version="v2",
):
    if chunk["type"] == "messages":
        msg, _ = chunk["data"]
        if isinstance(msg, AIMessageChunk) and msg.content:
            display_streaming_content(msg.content)

    elif chunk["type"] == "values" and chunk.get("interrupts"):
        interrupt_info = chunk["interrupts"][0].value
        user_response = get_user_input(interrupt_info)
        initial_input = Command(resume=user_response)
        break

    elif chunk["type"] == "updates":
        current_node = list(chunk["data"].keys())[0]
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    saw_interrupt = False
    for chunk in graph.stream(
        {},
        stream_mode=["values"],
        config=config,
        version="v2",
    ):
        if chunk["type"] == "values" and chunk.get("interrupts"):
            saw_interrupt = True
            break
    assert saw_interrupt
    print("✓ langgraph-interrupts-hitl-stream")
# :remove-end:

File: src/code-samples/langchain/langgraph-interrupts-hitl-stream.ts
// :remove-start:
import {
  Annotation,
  Command,
  END,
  MemorySaver,
  START,
  StateGraph,
  interrupt,
} from "@langchain/langgraph";

const StreamState = Annotation.Root({
  step: Annotation<number>(),
  done: Annotation<boolean>(),
});

function firstInterrupt(_state: typeof StreamState.State) {
  interrupt("first question");
  return { step: 1 };
}

function secondInterrupt(_state: typeof StreamState.State) {
  interrupt("second question");
  return { done: true };
}

const graph = new StateGraph(StreamState)
  .addNode("first", firstInterrupt)
  .addNode("second", secondInterrupt)
  .addEdge(START, "first")
  .addEdge("first", "second")
  .addEdge("second", END)
  .compile({ checkpointer: new MemorySaver() });

const config = { configurable: { thread_id: "stream-1" } };
const initialInput: Record<string, unknown> = {};

function displayStreamingContent(_content: string): void {
  // Application-specific UI hook
}

async function getUserInput(_interruptInfo: unknown): Promise<string> {
  return "ok";
}
// :remove-end:

// :snippet-start: langgraph-interrupts-hitl-stream-js
import { Command } from "@langchain/langgraph";

let streamInput: Record<string, unknown> | Command = initialInput;

while (true) {
  const stream = await graph.streamEvents(streamInput, {
    ...config,
    version: "v3",
  });

  // Stream LLM message chunks (including any in subgraphs) as they arrive.
  for await (const message of stream.messages) {
    for await (const token of message.text) {
      displayStreamingContent(token);
    }
  }

  // After the run finishes (or pauses), check for interrupts and resume.
  if (!stream.interrupted) {
    const finalState = await stream.output;
    break;
  }

  const interruptInfo = stream.interrupts[0].value;
  const userResponse = await getUserInput(interruptInfo);
  streamInput = new Command({ resume: userResponse });
}
// :snippet-end:

// :remove-start:
async function main() {
  const testConfig = { configurable: { thread_id: "stream-test" } };
  let streamInput: Record<string, unknown> | Command = {};
  let resumeRounds = 0;
  let final: typeof StreamState.State | undefined;

  while (true) {
    const testStream = await graph.streamEvents(streamInput, {
      ...testConfig,
      version: "v3",
    });
    for await (const _message of testStream.messages) {
      // Drain message stream so the run completes.
    }
    if (!testStream.interrupted) {
      final = await testStream.output;
      break;
    }
    streamInput = new Command({ resume: "ok" });
    resumeRounds += 1;
  }

  if (resumeRounds !== 2) {
    throw new Error(`Expected 2 resume rounds, got ${resumeRounds}`);
  }
  if (!final?.done) {
    throw new Error(`Expected done=true, got ${JSON.stringify(final)}`);
  }
  console.log("✓ langgraph-interrupts-hitl-stream");
}

main();
// :remove-end:
